File: Day2/Day2.py
import re

def half_1(inp):
    ranges = inp.split(',')
    IdSum = 0

    for r in ranges:
        startString,endString = r.split('-')
        start = int(startString)
        end = int(endString)

        currentNum = start

        if(len(startString) % 2 != 0):
            if(len(endString) == len(startString)):
                continue
            
            currentNum = pow(10, len(startString))

        while(currentNum <= end):
            curNumString = str(currentNum)
            first = int(curNumString[:len(curNumString)//2])
            sec = int(curNumString[len(curNumString)//2:])

            if(len(str(sec)) != len(curNumString)//2):
                sec = pow(10, len(curNumString) // 2 - 1)
                currentNum = int(str(first) + str(sec)) 
                continue

            if(first == sec):
                IdSum += int(str(first) + str(sec)) 

            if(first > sec):
                sec += 1
            elif(first < sec):
                if(int(str(first) + str(first)) > start):
                    sec = first
                else:
                    first += 1
            else:
                first += 1
                sec += 1
            
            currentNum = int(str(first) + str(sec)) 
        
    return IdSum

def half_2(inp):
    ranges = inp.split(',')
    IdSum = 0

    for r in ranges:
        startString,endString = r.split('-')
        start = int(startString)
        end = int(endString)

        reg = r"(.+)\1+"
        currentNum = start

        while(currentNum <= end):
            m = re.fullmatch(reg, str(currentNum))
            if(m):
                IdSum += int(m[0])
            
            currentNum += 1

    return IdSum


inpFile = open("input.txt", "r")
inp = inpFile.read()

print("Answer to Part 1:", half_1(inp))
print()
print("Answer to Part 2:", half_2(inp))

File: Day2/Day2.cpp

#include <iostream>
#include <fstream>
#include <string>
#include <sstream>
#include <vector>
#include <regex>

using namespace std;

const unsigned long powArray[12]={0, 10, 100, 1000, 10000, 100000, 1000000, 10000000, 100000000, 1000000000, 10000000000, 100000000000};

unsigned long part_1(vector<string>& ranges){

    unsigned long IdSum = 0;

    for(string r : ranges){
        string startString = r.substr(0,r.find('-'));
        string endString = r.substr(r.find('-') + 1, r.npos);
        unsigned long start = stol(startString);
        unsigned long end = stol(endString);

        unsigned long curNum = start;

        if(startString.length() % 2 !=0){
            if(endString.length() == startString.length()){
                continue;
            }
            curNum = powArray[startString.length()];
        }

        while(curNum <= end){
            string curNumString = to_string(curNum);
            unsigned long first = stol(curNumString.substr(0, curNumString.length() / 2));
            unsigned long sec = stol(curNumString.substr(curNumString.length() / 2, curNumString.npos));
            
            if(to_string(sec).length() != curNumString.length() / 2){
                sec = powArray[curNumString.length() / 2 - 1];
                curNum = stol(to_string(first) + to_string(sec));
                continue;
            }

            if(first == sec){
                IdSum = IdSum + curNum;
            }

            if(first > sec){
                sec++;
                if(stol(to_string(first) + to_string(first)) > start){
                    sec = first;
                }
                else{
                    sec++;
                }
                
            }
            else if(first < sec){
                if(stol(to_string(first) + to_string(first)) > start){
                    sec = first;
                }
                else{
                    first++;
                }
            }else{
                first++;
                sec++;
            }
            curNum = stol(to_string(first) + to_string(sec));
        }
    }

    return IdSum;
}

unsigned long part_2(vector<string>& ranges){

    unsigned long IdSum = 0;

    for(string r : ranges){
        int mid = r.find('-');
        string startString = r.substr(0,mid);
        string endString = r.substr(mid + 1, r.npos);
        unsigned long end = stol(endString);

        unsigned long curNum = stol(startString);
        regex reg("(.+)\\1+");

        while(curNum <= end){
            string curNumString = to_string(curNum);
            if(regex_match(curNumString, reg)){
                IdSum = IdSum + curNum;
            }
            curNum++;
        }
    }
    return IdSum;
}

int main(){

    ifstream f("input.txt");
    
    if(!f.is_open()){
        cout << "file read error" << endl;
        return 1;
    }

    string inpString;
    vector<string> ranges;

    while(getline(f, inpString)){
        string temp;
        stringstream ss(inpString);

        while(getline(ss, temp, ',')){
            ranges.push_back(temp);
        }
    }

    cout << "Answer to Part 1: " << part_1(ranges) << endl;
    cout << "Answer to Part 2: " << part_2(ranges) << endl;
    return 0;
}